<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [606]:
import pandas as pd
import psycopg2

In [607]:
# вставьте сюда параметры подключения. 
DBNAME = 
USER = 
PASSWORD = 
HOST = 
PORT = 

In [608]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [609]:
# текст запроса
query_3_1 = f'''
select count(*)
from vacancies
'''

In [610]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [611]:
# текст запроса
query_3_2 = f'''
select count(*)
from employers
'''

In [612]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [613]:
# текст запроса
query_3_3 = f'''
select count(*)
from areas
'''

In [614]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [615]:
# текст запроса
query_3_4 = f'''
select count(*)
from industries
'''

In [616]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

In [617]:
# выводы по предварительному анализу данных

**Количество вакансий:**  
В базе данных содержится большое количество вакансий - **49 197**. Это свидетельствует о разнообразии предлагаемых рабочих мест и возможностях для соискателей.

**Количество работодателей:**  
Существует **23 501** работодатель, что указывает на то, что множество организаций предлагают рабочие вакансии в данной базе данных. 

**Количество регионов:** 
База данных содержит информацию о вакансиях в **1 362** различных регионах. Это означает, что вакансии доступны в различных географических областях, что может быть важным для соискателей, которые ищут работу в конкретном регионе.

**Количество сфер деятельности:**  
Существует **294** разных сферы деятельности в базе данных. Это указывает на разнообразие отраслей, в которых доступны рабочие места, и может предоставить соискателям возможность выбора вакансий, соответствующих их навыкам и интересам.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [618]:
# текст запроса
query_4_1 = f'''
select a.name, count(v.id) cnt 
from vacancies v join areas a on v.area_id = a.id
group by 1
order by 2 desc
'''

In [619]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [620]:
# текст запроса
query_4_2 = f'''
select count(*)
from vacancies
where salary_from is not null or salary_to is not null
'''

In [621]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [622]:
# текст запроса
query_4_3 = f'''
select avg(salary_from), avg(salary_to)
from vacancies
'''

In [623]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg            avg
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [624]:
# текст запроса
query_4_4 = f'''
select schedule, employment, count(*) vacancy_count
from vacancies
group by 1, 2
order by 3 desc
'''

In [625]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancy_count
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [626]:
# текст запроса
query_4_5 = f'''
select experience, count(*) vacancy_count
from vacancies
group by 1
order by 2 
'''

In [627]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancy_count
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

In [628]:
# выводы по детальному анализу вакансий

**Регионы с наибольшим количеством вакансий:**  
Москва, Санкт-Петербург, Минск, Новосибирск и Алматы являются лидерами по количеству предложений о работе и, следовательно, предоставляют больше возможностей для соискателей.

**Количество вакансий с заполненной информацией о зарплате:**  
**24 073**. Это означает, что многие работодатели предоставляют информацию о зарплате, что может быть полезно для соискателей при выборе вакансий.

**Зарплатные вилки:**  
Нижняя граница зарплатной вилки составляет **71 065**, а верхняя - **11 537**. Эти значения могут служить ориентиром для соискателей при оценке предложений о зарплате.

**Сочетания типа рабочего графика и типа трудоустройства:**  
Самые популярные сочетания - **"Полный день + Полная занятость" с 35 367 вакансиями** и **"Удаленная работа + Полная занятость" с 7 802 вакансиями**. Эти данные могут помочь соискателям оценить доступные варианты работы.

**Требуемый опыт работы:**  
Наибольшее количество вакансий требует опыт **от 1 года до 3 лет**. Это важная информация для соискателей, чтобы определить, какой опыт требуется для большинства доступных вакансий.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [629]:
# текст запроса
query_5_1 = f'''
select e.name, count(v.employer_id) vacancy_count
from vacancies v join employers e on v.employer_id = e.id
group by 1
order by 2 desc
limit 5
'''

In [630]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  vacancy_count
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [631]:
# текст запроса
query_5_2 = f'''
select a.name, count(e.id) employers_count, count(v.id) vacancy_count
from areas a 
left join employers e on e.area = a.id
left join vacancies v on v.area_id = a.id
where v.id is null
group by 1
order by 2 desc
'''

In [632]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  employers_count  vacancy_count
0                            Россия              410              0
1                         Казахстан              207              0
2                Московская область               75              0
3                Краснодарский край               19              0
4                          Беларусь               18              0
..                              ...              ...            ...
588                        Кувандык                0              0
589                        Куйбышев                0              0
590  Куйбышево (Ростовская область)                0              0
591                        Култаево                0              0
592                        Кульсары                0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [633]:
# текст запроса
query_5_3 = f'''
select e.name, count(distinct v.area_id) areas_count
from employers e
join vacancies v on e.id = v.employer_id
group by 1
order by 2 desc
'''

In [634]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  areas_count
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
14761                  UniSol            1
14762            UNISTORY LLC            1
14763                   UNIT6            1
14764     United Distribution            1
14765              UNITEL LLC            1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [635]:
# текст запроса
query_5_4 = f'''
select count(*) employer_count
from employers e
left join employers_industries ei on e.id = ei.employer_id
where ei.industry_id is null
'''

In [636]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employer_count
0            8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [637]:
# текст запроса
query_5_5 = f'''
select e.name 
from employers e
left join employers_industries ei on e.id = ei.employer_id
group by 1
having count(ei.industry_id)=4
order by 1
limit 3
'''

In [638]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  101 Интернет
1      21vek.by
2          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [639]:
# текст запроса
query_5_6 = f'''
select count(e.name)
from employers_industries ei
join employers e on e.id = ei.employer_id
join industries i on i.id = ei.industry_id
where i.name like 'Разработка программного обеспечения'
'''

In [640]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [641]:
# код для получения списка городов-милионников
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = tuple(pd.read_html(url)[1]['Город'])

In [642]:
# текст запроса
query_5_7 = f'''
with t_1 as (
    select a.name, count(v.id) vacancy_count
    from employers as e
    join vacancies as v on e.id = v.employer_id
    join areas as a on v.area_id = a.id
    where e.name = 'Яндекс'
    group by a.name
    having a.name in {cities}
    order by 2 asc
),
total_count as (
    select sum(vacancy_count) total_vacancy_count
    from t_1
)
select name, vacancy_count
from t_1
union all
select 'Total', total_vacancy_count
from total_count
'''

In [643]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  vacancy_count
0              Омск           21.0
1         Челябинск           22.0
2        Красноярск           23.0
3         Волгоград           24.0
4             Пермь           25.0
5            Казань           25.0
6    Ростов-на-Дону           25.0
7               Уфа           26.0
8            Самара           26.0
9         Краснодар           30.0
10          Воронеж           32.0
11      Новосибирск           35.0
12  Нижний Новгород           36.0
13     Екатеринбург           39.0
14  Санкт-Петербург           42.0
15           Москва           54.0
16            Total          485.0

***

In [644]:
# выводы по анализу работодателей

**Топ-5 работодателей по количеству вакансий:**  
- Яндекс с 1 933 вакансиями;
- Ростелеком с 491 вакансией;
- Тинькофф с 444 вакансиями;
- СБЕР с 428 вакансиями;
- Газпром нефть с 331 вакансией.

**Регионы, в которых нет вакансий:**  
В регионе "Россия" отсутствуют вакансии, и здесь находится наибольшее количество работодателей (410). Это может быть связано с тем, что некоторые работодатели могут предлагать вакансии на уровне страны, без привязки к конкретному региону.

**Работодатель с наибольшим охватом регионов:**  
Яндекс с 181 регионом. Это означает, что Яндекс публикует вакансии в большом количестве различных местоположений, что может быть интересным для соискателей, ищущих работу в разных регионах. Также компания Яндекс имеет 485 вакансий в городах-миллионниках, причем большинство из них находятся в Москве (54) и Санкт-Петербурге (42).

**Работодатели без указания сферы деятельности:**  
Существует 8 419 работодателей, у которых не указана сфера деятельности. Это может означать, что некоторые работодатели не предоставили информацию о своей деятельности, что может затруднить поиск работы для соискателей. А вот сфера деятельности "Разработка программного обеспечения" указана у 3 553 работодателей. Это свидетельствует о высоком спросе на специалистов в этой области.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [645]:
# текст запроса
query_6_1 = f'''
select count(*) vacancy_count
from vacancies
where name ilike '%data%' or name ilike '%данн%'
'''

In [646]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


vacancy_count
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [647]:
# текст запроса
query_6_2 = f'''
select count(*)
from vacancies
where ((name ilike '%data scientist%') or (name ilike '%data science%') or (name ilike '%исследователь данных%')
or (name like '%ML%' and name not like '%HTML%')
or (name ilike '%machine learning%')
or (name ilike '%машинн%обучен%')
and (key_skills is not null))
and (name ilike '%junior%' or experience ilike '%Нет опыта%' or employment ilike '%Стажировка%')
'''

In [648]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [649]:
# текст запроса
query_6_3 = f'''
select count(*)
from vacancies
where ((name ilike '%data scientist%') or (name ilike '%data science%') or (name ilike '%исследователь данных%')
or (name like '%ML%' and name not like '%HTML%')
or (name ilike '%machine learning%')
or (name ilike '%машинн%обучен%')
and (key_skills is not null))
and (key_skills ilike '%SQL%' or key_skills ilike '%postgres%')
'''

In [650]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [651]:
# текст запроса
query_6_4 = f'''
select count(*)
from vacancies
where ((name ilike '%data scientist%') or (name ilike '%data science%') or (name ilike '%исследователь данных%')
or (name like '%ML%' and name not like '%HTML%')
or (name ilike '%machine learning%')
or (name ilike '%машинн%обучен%')
and (key_skills is not null))
and (key_skills ilike '%Python%')
'''

In [652]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [653]:
# текст запроса
query_6_5 = f'''
select round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2)
from vacancies
where ((name ilike '%data scientist%') or (name ilike '%data science%') or (name ilike '%исследователь данных%')
or (name like '%ML%' and name not like '%HTML%')
or (name ilike '%machine learning%')
or (name ilike '%машинн%обучен%')
and (key_skills is not null))
'''

In [654]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [655]:
# текст запроса
query_6_6 = f'''
select 
    distinct(experience), 
    round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to))) wage
from vacancies
where ((name ilike '%data scientist%') or (name ilike '%data science%') or (name ilike '%исследователь данных%')
or (name like '%ML%' and name not like '%HTML%')
or (name ilike '%machine learning%')
or (name ilike '%машинн%обучен%')
and (key_skills is not null))
and (salary_from is not null or salary_to is not null)
group by 1
'''

In [656]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience      wage
0           Нет опыта   74643.0
1  От 1 года до 3 лет  145236.0
2       От 3 до 6 лет  243115.0

***

In [657]:
# выводы по предметному анализу

- Всего существует 1771 вакансия, связанных с данными, что указывает на **активный спрос** на специалистов в этой области.

- 51 вакансия предназначена для начинающих дата-сайентистов, что свидетельствует о **наличии возможностей для новичков** в этой области.

- Из 201 вакансии, где ключевыми навыками являются SQL или PostgreSQL, можно сделать вывод, что **знание SQL является важным требованием для работы в области data science**.

- 351 вакансия требует знания Python, что подчеркивает **популярность этого языка программирования** в анализе данных и науке о данных.

- В среднем, вакансии для дата-сайентистов требуют 6-7 ключевых навыков, что указывает на **необходимость обширных компетенций в этой профессии**.

- Зарплаты для дата-сайентистов значительно различаются в зависимости от опыта. Новички могут ожидать зарплату в размере 74 643 рублей, в то время как опытные специалисты с опытом от 3 до 6 лет могут получать в среднем 243 115 рублей. Это подчеркивает **важность опыта влияющего на уровень заработной платы** в этой области.

# Общий вывод по проекту

In [658]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

### Количество вакансий
В базе данных содержится значительное количество вакансий - **49 197**. Это указывает на высокую активность на рынке труда и множество возможностей для соискателей.

### Разнообразие рынка труда
Существует большое количество работодателей (**23 501**) и различных регионов (**1 362**), что свидетельствует о разнообразии предложений и географических возможностях для соискателей.

### Распределение по регионам
Москва, Санкт-Петербург, Минск, Новосибирск и Алматы являются лидерами по количеству вакансий, что предоставляет больше возможностей для соискателей в этих городах. Большое количество вакансий в данных регионах можно объяснить их статусом экономических и технологических центров, наличием ведущих университетов, развитой инфраструктурой и бизнес-средой, что привлекает компании и специалистов.

### Специализация
В базе данных представлены **294** разных сферы деятельности, что говорит о разнообразии отраслей и множестве предложений.

### Информация о зарплате
Значительное количество вакансий (**24 073**) предоставляют информацию о зарплате, что полезно для соискателей при выборе вакансий. Зарплатные вилки варьируются от **71 065** до **11 537**, что может служить ориентиром для оценки предложений о зарплате.

### Типы рабочего графика и трудоустройства
Сочетания **"Полный день + Полная занятость"** и **"Удаленная работа + Полная занятость"** являются наиболее популярными, что предоставляет разнообразные варианты для соискателей. Эти сочетания могут быть популярными, потому что они предоставляют стабильную занятость и гибкость в работе, соответствуя ожиданиям как работодателей, так и работников.

### Требуемый опыт работы
Большинство вакансий требуют опыта **от 1 года до 3 лет**, что важно для соискателей при оценке своей пригодности к вакансии.

### Вакансии в области данных
Существует активный спрос на специалистов в области data science. Знание SQL и Python является ключевыми навыками в этой области. Вакансии для дата-сайентистов требуют обширных компетенций, и опыт работы существенно влияет на уровень заработной платы.

Рассмотренные мной данные предоставляют ценную информацию для соискателей, работодателей и аналитиков рынка труда. Они позволяют понять тенденции, определить актуальные навыки и оценить рыночную конкуренцию, что может быть полезным как для поиска работы, так и для планирования рекрутинговых стратегий.

**В дополнительных исследованиях можно было бы рассмотреть следующие аспекты:**
- Сравнение средних зарплат в разных регионах и выявление факторов, влияющих на различия в заработной плате.
- Анализ связи между уровнем опыта и уровнем заработной платы, чтобы определить, как опыт влияет на зарплату в разных отраслях.

In [659]:
# текст запроса
query_7_1 = f'''
select
    a.name region,
    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) wage
from vacancies v 
join areas a on v.area_id = a.id
where (salary_from is not null or salary_to is not null)
group by 1
order by 2 desc
limit 16
'''

In [660]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\2931715827.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


region      wage
0                             Германия  452751.0
1                             Болгария  324348.0
2                              Испания  300000.0
3                           Черногория  288589.0
4                                Литва  263019.0
5                               Польша  257156.0
6   Петропавловка (Республика Бурятия)  250000.0
7                               Турция  235435.0
8                                 Кипр  223736.0
9                              Кутаиси  220475.0
10                Дзержинск (Беларусь)  220349.0
11                             Армения  219570.0
12                         Таджикистан  215094.0
13                      Великобритания  208962.0
14                              Сербия  208423.0
15                             Тбилиси  207156.0

In [661]:
# текст запроса
query_7_1_2 = f'''
select
    a.name region,
    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) wage
from vacancies v 
join areas a on v.area_id = a.id
where (salary_from is not null or salary_to is not null)
group by 1
order by 2 asc
limit 16
'''

In [662]:
# результат запроса
df = pd.read_sql_query(query_7_1_2, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\1442532564.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1_2, connection)


region     wage
0                     Шахтинск   9764.0
1                  Степногорск  12205.0
2                       Кобрин  12389.0
3                           Ош  12743.0
4                        Урень  13800.0
5                   Рассказово  13890.0
6   Красное (Липецкая область)  14500.0
7             Гусь-Хрустальный  15000.0
8                      Кимовск  15445.0
9                 Отеген-Батыр  15622.0
10                    Темиртау  16171.0
11                       Тараз  16476.0
12                      Рудный  17636.0
13                  Слободской  18000.0
14                      Алушта  18767.0
15                       Семей  19833.0

#### Исходя из сравнения средних зарплат в разных регионах, можно сделать следующие выводы:

**Самые высокие зарплаты:**  
- Германия имеет наивысшие средние зарплаты в рассматриваемых регионах, что связано с высоким экономическим развитием и квалификацией трудовых ресурсов.
- Средние зарплаты в других европейских странах, таких как Болгария, Испания, Литва и Польша, также высоки и могут быть связаны с развитой экономикой и     доступностью квалифицированной рабочей силы.
- Некоторые регионы внутри стран, такие как Петропавловка (Республика Бурятия) и Дзержинск (Беларусь), также имеют высокие зарплаты, возможно, благодаря специфическим отраслям или инвестициям.
- Великобритания, Сербия и Тбилиси также показывают высокие средние зарплаты, вероятно, из-за развитой экономики и бизнес-сферы.

**Самые низкие зарплаты:**  
- Регионы с самыми низкими средними зарплатами, такие как Шахтинск, Степногорск и Кобрин, могут сталкиваться с экономическими и социальными вызовами, что влияет на уровень заработных плат.
- Ош, Темиртау, и Тараз также имеют низкие средние зарплаты, что может быть связано с местными экономическими условиями и структурой занятости.
- В регионах, где средние зарплаты ниже, могут сказываться факторы, такие как низкая инвестиционная активность, отсутствие развитых отраслей, высокая безработица и другие социально-экономические факторы.

В целом, средние зарплаты сильно зависят от экономических и социальных условий в регионе, структуры экономики, наличия инвестиций и уровня квалификации трудовых ресурсов. Дополнительный анализ данных и факторов каждого региона может дать более подробное понимание различий в заработной плате.

In [673]:
# текст запроса
query_7_2 = f'''
select 
    i.name industy,
    v.experience, 
    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) wage
from vacancies v
join employers_industries ei on v.employer_id = ei.employer_id
join industries i on ei.industry_id = i.id
where (salary_from is not null or salary_to is not null)
group by 1, 2
order by 3 desc
limit 16
'''

In [674]:
# результат запроса
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\746169716.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


industy     experience      wage
0   Бытовая химия, парфюмерия, косметика  (произво...    Более 6 лет  500000.0
1           Агрохимия (продвижение, оптовая торговля)    Более 6 лет  500000.0
2                            Агрохимия (производство)    Более 6 лет  500000.0
3   Бытовая химия, парфюмерия, косметика  (продвиж...    Более 6 лет  462273.0
4                            Клинические исследования    Более 6 лет  400000.0
5                Лаборатория, исследовательский центр    Более 6 лет  400000.0
6   Розничная сеть (книги, игрушки, подарки, часы ...    Более 6 лет  305118.0
7                                  Кадровые агентства    Более 6 лет  302154.0
8   Металлические изделия, металлоконструкции (про...    Более 6 лет  300000.0
9                                  Интернет-провайдер    Более 6 лет  300000.0
10  Продукция черной металлургии (продвижение, опт...    Более 6 лет  300000.0
11  Интернет-компания (поисковики, платежные систе...    Более 6 лет  268118.0
12                                     Игорный бизнес  От 3 до 6 лет  252848.0
13  Продукция цветной металлургии (продвижение, оп...    Более 6 лет  237500.0
14                           Юридическое обслуживание    Более 6 лет  237500.0
15  Аудит, управленческий учет, финансово-юридичес...    Более 6 лет  225000.0

In [675]:
# текст запроса
query_7_2_2 = f'''
select 
    i.name industy,
    v.experience, 
    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) wage
from vacancies v
join employers_industries ei on v.employer_id = ei.employer_id
join industries i on ei.industry_id = i.id
where (salary_from is not null or salary_to is not null)
group by 1, 2
order by 3 asc
limit 16
'''

In [676]:
# результат запроса
df = pd.read_sql_query(query_7_2_2, connection)
df

C:\Users\Polina\AppData\Local\Temp\ipykernel_13396\619807298.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2_2, connection)


industy experience     wage
0                               Табак  (производство)  Нет опыта  13254.0
1                          ВЭД, таможенное оформление  Нет опыта  15000.0
2     Обеспечение пожарной безопасности, молниезащиты  Нет опыта  18307.0
3   Оборудование и станки для металлургии и металл...  Нет опыта  19500.0
4                      Детское питание (производство)  Нет опыта  20137.0
5                                 Вино (производство)  Нет опыта  22000.0
6                                 Спортивное обучение  Нет опыта  22500.0
7              Табак  (продвижение, оптовая торговля)  Нет опыта  22877.0
8                                     Энергоснабжение  Нет опыта  23189.0
9   Производство и продажа рекламно-сувенирной про...  Нет опыта  23603.0
10                        Обучение иностранным языкам  Нет опыта  24234.0
11  Обучение искусствам (рисование, пение, танцы, ...  Нет опыта  24409.0
12                                       Судостроение  Нет опыта  25000.0
13                Архив, библиотека, искусствоведение  Нет опыта  25454.0
14  Контроль качества, экспертиза, испытания и сер...  Нет опыта  27176.0
15  Интернет-компания (поисковики, платежные систе...  Нет опыта  27626.0

#### Исходя из сравнения средних зарплат в разных отраслях и с разным уровнем опыта, можно сделать следующие выводы:

**Самые высокие зарплаты:**  
- В отраслях, таких как "Бытовая химия, парфюмерия, косметика" и "Агрохимия," средние зарплаты значительно выше, особенно для специалистов с более чем 6 годами опыта. Это может быть связано с высокой специализацией и требовательностью в этих областях.
- Другие отрасли, такие как "Клинические исследования" и "Лаборатория, исследовательский центр," также предлагают высокие зарплаты для опытных специалистов, что может быть обусловлено высокой важностью и квалификацией работников в этих областях.
- "Интернет-провайдер" и "Интернет-компания" также входят в список с высокими зарплатами, вероятно, из-за динамичного характера работы и высокой конкуренции на рынке IT.

**Самые низкие зарплаты:**  
- В отраслях, где нет требования к опыту, средние зарплаты значительно ниже. Это может быть связано с тем, что в этих отраслях требуется менее специфическая квалификация и опыт.
- "Спортивное обучение," "Обучение иностранным языкам," и "Обучение искусствам" также входят в список с низкими зарплатами для начинающих специалистов, возможно, из-за специфики образовательных сфер.
- "Интернет-компания" также присутствует с низкими зарплатами для специалистов без опыта, что может быть связано с конкурентной средой в IT-сфере и большим числом начинающих специалистов.

В целом, уровень зарплаты сильно зависит от отрасли и уровня опыта. Специфика отрасли и конкурентоспособность могут влиять на уровень заработной платы, а также на важность квалификации и опыта.